File: ios/KetoKasse/Features/Week/WeekPlanner.swift
import Foundation

enum PlanWeekday: Int, CaseIterable, Identifiable, Codable, Sendable {
    case monday = 1, tuesday, wednesday, thursday, friday, saturday, sunday

    var id: Int { rawValue }

    var shortLabel: String {
        switch self {
        case .monday: "Man"
        case .tuesday: "Tir"
        case .wednesday: "Ons"
        case .thursday: "Tor"
        case .friday: "Fre"
        case .saturday: "Lør"
        case .sunday: "Søn"
        }
    }

    static func on(_ date: Date, calendar: Calendar = .current) -> PlanWeekday {
        let sundayIndexed = calendar.component(.weekday, from: date)
        let mondayIndexed = (sundayIndexed + 5) % 7 + 1
        return PlanWeekday(rawValue: mondayIndexed)!
    }
}

struct WeekPlan: Equatable, Codable, Sendable {
    static let dayCount = 7
    static let dinnerCount = 5

    var seed: UInt64
    var slots: [String?]

    var filledTitles: [String] {
        slots.compactMap { $0 }
    }

    var filledCount: Int { filledTitles.count }

    func title(on day: PlanWeekday) -> String? {
        slots[day.rawValue - 1]
    }

    func dish(on day: PlanWeekday) -> Dish? {
        title(on: day).flatMap(DishPool.dish(titled:))
    }
}

struct SplitMix64: RandomNumberGenerator {
    var state: UInt64

    init(seed: UInt64) {
        state = seed
    }

    mutating func next() -> UInt64 {
        state &+= 0x9E3779B97F4A7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58476D1CE4E5B9
        z = (z ^ (z >> 27)) &* 0x94D049BB133111EB
        return z ^ (z >> 31)
    }
}

enum WeekPlanner {
    static func isoWeekSeed(for date: Date, calendar: Calendar = .current) -> UInt64 {
        var iso = Calendar(identifier: .iso8601)
        iso.timeZone = calendar.timeZone
        let week = iso.component(.weekOfYear, from: date)
        let year = iso.component(.yearForWeekOfYear, from: date)
        return UInt64(year) &* 100 &+ UInt64(week)
    }

    static func isoWeekKey(for date: Date, calendar: Calendar = .current) -> String {
        var iso = Calendar(identifier: .iso8601)
        iso.timeZone = calendar.timeZone
        let week = iso.component(.weekOfYear, from: date)
        let year = iso.component(.yearForWeekOfYear, from: date)
        return "\(year)-W\(week)"
    }

    static func mix(weekSeed: UInt64, generation: UInt64) -> UInt64 {
        weekSeed &* 1_000_003 &+ generation
    }

    static func generate(
        seed: UInt64,
        avoiding previousTitles: Set<String>,
        pool: [Dish] = DishPool.all
    ) -> WeekPlan {
        var rng = SplitMix64(seed: seed)
        let dishes = pickDishes(pool: pool, avoiding: previousTitles, rng: &rng)
        let slots = place(dishes, rng: &rng)
        return WeekPlan(seed: seed, slots: slots.map { $0?.title })
    }

    static func proteinsAreValid(_ slots: [Dish?]) -> Bool {
        guard slots.count == WeekPlan.dayCount else { return false }
        guard slots.compactMap({ $0 }).count == WeekPlan.dinnerCount else { return false }
        for index in 0..<(WeekPlan.dayCount - 1) {
            guard let a = slots[index], let b = slots[index + 1] else { continue }
            if a.protein == b.protein {
                return false
            }
        }
        return true
    }

    private static func pickDishes(
        pool: [Dish],
        avoiding previousTitles: Set<String>,
        rng: inout SplitMix64
    ) -> [Dish] {
        let pizza = pool.first { $0.title == DishPool.pizzaTitle } ?? DishPool.pizza
        let rest = pool.filter { $0.title != pizza.title }
        let fresh = shuffled(rest.filter { !previousTitles.contains($0.title) }, rng: &rng)
        let stale = shuffled(rest.filter { previousTitles.contains($0.title) }, rng: &rng)
        let ranked = fresh + stale

        var four: [Dish] = []
        var chickenCount = 1
        for dish in ranked where four.count < 4 {
            if dish.protein == pizza.protein && chickenCount >= 4 {
                continue
            }
            four.append(dish)
            if dish.protein == pizza.protein {
                chickenCount += 1
            }
        }
        for dish in ranked where four.count < 4 {
            if !four.contains(dish) {
                four.append(dish)
            }
        }
        return [pizza] + four
    }

    private static func place(_ dishes: [Dish], rng: inout SplitMix64) -> [Dish?] {
        let empties = shuffled(emptyPairs(), rng: &rng)
        for _ in 0..<200 {
            let perm = shuffled(dishes, rng: &rng)
            for empty in empties {
                let slots = build(perm, empty: empty)
                if proteinsAreValid(slots) {
                    return slots
                }
            }
        }
        for perm in permutations(dishes) {
            for empty in emptyPairs() {
                let slots = build(perm, empty: empty)
                if proteinsAreValid(slots) {
                    return slots
                }
            }
        }
        return build(dishes, empty: (2, 5))
    }

    private static func build(_ dishes: [Dish], empty: (Int, Int)) -> [Dish?] {
        var slots: [Dish?] = Array(repeating: nil, count: WeekPlan.dayCount)
        var dishIndex = 0
        for day in 0..<WeekPlan.dayCount {
            if day == empty.0 || day == empty.1 { continue }
            slots[day] = dishes[dishIndex]
            dishIndex += 1
        }
        return slots
    }

    private static func emptyPairs() -> [(Int, Int)] {
        var pairs: [(Int, Int)] = []
        for a in 0..<WeekPlan.dayCount {
            for b in (a + 1)..<WeekPlan.dayCount {
                pairs.append((a, b))
            }
        }
        return pairs
    }

    private static func shuffled<T>(_ items: [T], rng: inout SplitMix64) -> [T] {
        var items = items
        var i = items.count
        while i > 1 {
            i -= 1
            let j = Int(rng.next() % UInt64(i + 1))
            items.swapAt(i, j)
        }
        return items
    }

    private static func permutations<T>(_ items: [T]) -> [[T]] {
        var items = items
        var result: [[T]] = []
        permute(&items, start: 0, into: &result)
        return result
    }

    private static func permute<T>(_ items: inout [T], start: Int, into result: inout [[T]]) {
        if start == items.count {
            result.append(items)
            return
        }
        for index in start..<items.count {
            items.swapAt(start, index)
            permute(&items, start: start + 1, into: &result)
            items.swapAt(start, index)
        }
    }
}

File: ios/scripts/week_planner.py
from __future__ import annotations

import re
from dataclasses import dataclass
from itertools import combinations, permutations
from pathlib import Path

MASK = (1 << 64) - 1
PIZZA_TITLE = "Kyllingpizza med mozzarella og basilikum"
DAY_COUNT = 7
DINNER_COUNT = 5


@dataclass(frozen=True)
class Dish:
    title: str
    protein: str


class SplitMix64:
    def __init__(self, state: int) -> None:
        self.state = state & MASK

    def next(self) -> int:
        self.state = (self.state + 0x9E3779B97F4A7C15) & MASK
        z = self.state
        z = ((z ^ (z >> 30)) * 0xBF58476D1CE4E5B9) & MASK
        z = ((z ^ (z >> 27)) * 0x94D049BB133111EB) & MASK
        return (z ^ (z >> 31)) & MASK


def shuffled(items: list, rng: SplitMix64) -> list:
    items = list(items)
    i = len(items)
    while i > 1:
        i -= 1
        j = rng.next() % (i + 1)
        items[i], items[j] = items[j], items[i]
    return items


def parse_pool(swift_text: str) -> list[Dish]:
    dishes = [
        Dish(title=title, protein=protein)
        for title, protein in re.findall(
            r'Dish\(title: "([^"]+)", protein: \.(\w+)\)',
            swift_text,
        )
    ]
    return dishes


def parse_pizza_title(swift_text: str) -> str:
    match = re.search(r'static let pizzaTitle = "([^"]+)"', swift_text)
    if not match:
        raise ValueError("pizzaTitle missing")
    return match.group(1)


def proteins_are_valid(slots: list[Dish | None]) -> bool:
    if len(slots) != DAY_COUNT:
        return False
    filled = [dish for dish in slots if dish is not None]
    if len(filled) != DINNER_COUNT:
        return False
    for index in range(DAY_COUNT - 1):
        left, right = slots[index], slots[index + 1]
        if left is not None and right is not None and left.protein == right.protein:
            return False
    return True


def empty_pairs() -> list[tuple[int, int]]:
    return list(combinations(range(DAY_COUNT), 2))


def build(dishes: list[Dish], empty: tuple[int, int]) -> list[Dish | None]:
    slots: list[Dish | None] = [None] * DAY_COUNT
    dish_index = 0
    for day in range(DAY_COUNT):
        if day in empty:
            continue
        slots[day] = dishes[dish_index]
        dish_index += 1
    return slots


def pick_dishes(pool: list[Dish], pizza_title: str, previous: set[str], rng: SplitMix64) -> list[Dish]:
    pizza = next(dish for dish in pool if dish.title == pizza_title)
    rest = [dish for dish in pool if dish.title != pizza.title]
    fresh = shuffled([dish for dish in rest if dish.title not in previous], rng)
    stale = shuffled([dish for dish in rest if dish.title in previous], rng)
    ranked = fresh + stale
    four: list[Dish] = []
    chicken_count = 1
    for dish in ranked:
        if len(four) == 4:
            break
        if dish.protein == pizza.protein and chicken_count >= 4:
            continue
        four.append(dish)
        if dish.protein == pizza.protein:
            chicken_count += 1
    for dish in ranked:
        if len(four) == 4:
            break
        if dish not in four:
            four.append(dish)
    return [pizza] + four


def place(dishes: list[Dish], rng: SplitMix64) -> list[Dish | None]:
    empties = shuffled(empty_pairs(), rng)
    for _ in range(200):
        perm = shuffled(dishes, rng)
        for empty in empties:
            slots = build(perm, empty)
            if proteins_are_valid(slots):
                return slots
    for perm in permutations(dishes):
        for empty in empty_pairs():
            slots = build(list(perm), empty)
            if proteins_are_valid(slots):
                return slots
    return build(dishes, (2, 5))


def generate(seed: int, previous: set[str], pool: list[Dish], pizza_title: str = PIZZA_TITLE) -> list[str | None]:
    rng = SplitMix64(seed)
    dishes = pick_dishes(pool, pizza_title, previous, rng)
    slots = place(dishes, rng)
    return [dish.title if dish else None for dish in slots]


def filled_titles(slots: list[str | None]) -> list[str]:
    return [title for title in slots if title]


def move_slot(slots: list[str | None], source: int, target: int) -> list[str | None]:
    if source == target:
        return list(slots)
    next_slots = list(slots)
    item = next_slots.pop(source)
    next_slots.insert(target, item)
    return next_slots


def assert_move_slot_behavior() -> None:
    if move_slot(["A", "B", "C", "D", "E", "F", "G"], 0, 3) != ["B", "C", "D", "A", "E", "F", "G"]:
        raise SystemExit("move down did not shift later days toward the source")
    if move_slot(["A", "B", "C", "D", "E", "F", "G"], 3, 0) != ["D", "A", "B", "C", "E", "F", "G"]:
        raise SystemExit("move up did not shift earlier days toward the source")
    if move_slot(["A", "B", None, "C", None, "D", "E"], 0, 2) != ["B", None, "A", "C", None, "D", "E"]:
        raise SystemExit("move onto Fri did not leave an empty source day")
    if move_slot(["A", "B", "C", "D", "E", "F", "G"], 2, 2) != ["A", "B", "C", "D", "E", "F", "G"]:
        raise SystemExit("same-slot move changed the plan")


def assert_week_pager_stays_free_until_drag(week_text: str) -> None:
    is_active = re.search(r"var isActive: Bool \{(?P<body>.*?)\n    \}", week_text, re.S)
    if is_active is None:
        raise SystemExit("WeekSlotDrag missing isActive")
    body = is_active.group("body")
    if "self != .inactive" in body or ".pressing, .dragging" in body or ".pressing,.dragging" in body:
        raise SystemExit("WeekSlotDrag.isActive treats press as a lift and locks the home pager")
    if ".dragging" not in body:
        raise SystemExit("WeekSlotDrag.isActive must be true only while dragging")
    if ".simultaneousGesture(slotGesture" not in week_text:
        raise SystemExit("slotGesture must be simultaneous so TabView can page")
    if re.search(r"\.gesture\(slotGesture", week_text):
        raise SystemExit("exclusive .gesture(slotGesture) blocks TabView paging")
    distance = re.search(r"DragGesture\(minimumDistance:\s*(\d+)", week_text)
    if distance is None:
        raise SystemExit("DragGesture must set minimumDistance")
    if int(distance.group(1)) < 10:
        raise SystemExit(
            f"DragGesture minimumDistance {distance.group(1)} is too low for page swipe"
        )


def assert_week_reorder_contract(week_text: str, store_text: str) -> None:
    banned_week = {
        "editMode": "WeekPlannerView still has editMode",
        ".onMove": "WeekPlannerView still has onMove",
        "draggable": "WeekPlannerView still has draggable",
        "dropDestination": "WeekPlannerView still has dropDestination",
        "draggableIfPresent": "WeekPlannerView still has draggableIfPresent",
        "Hold de tre strekene": "WeekPlannerView still has the three-bar copy",
        ".id(UUID())": "WeekPlannerView resets identity with UUID",
        "coordinateSpace: .local": "WeekPlannerView still drags in local tile space",
    }
    for needle, message in banned_week.items():
        if needle in week_text:
            raise SystemExit(message)
    if "List {" in week_text or "List(" in week_text:
        raise SystemExit("WeekPlannerView still has List")
    assert_week_pager_stays_free_until_drag(week_text)
    required_week = {
        "LongPressGesture": "WeekPlannerView missing LongPressGesture",
        "sequenced(before:": "WeekPlannerView missing sequenced drag",
        "DragGesture": "WeekPlannerView missing DragGesture",
        "@GestureState": "WeekPlannerView missing GestureState",
        "accessibilityReduceMotion": "WeekPlannerView missing Reduce Motion",
        "sensoryFeedback": "WeekPlannerView missing haptics",
        "coordinateSpace: .named": "WeekPlannerView must drag in a named board space",
        "WeekBoardDragActiveKey": "WeekPlannerView missing pager lock preference",
        "dropSettled": "WeekPlannerView must freeze the lift on drop",
    }
    for needle, message in required_week.items():
        if needle not in week_text:
            raise SystemExit(message)
    required_store = {
        "moveSlot": "WeekStore missing moveSlot",
        "remove(at:": "WeekStore missing remove(at:)",
        "insert(": "WeekStore missing insert",
    }
    for needle, message in required_store.items():
        if needle not in store_text:
            raise SystemExit(message)
    banned_store = {
        "fromOffsets": "WeekStore still has Array.move",
        "moveSlots": "WeekStore still has moveSlots",
        "swapAt": "WeekStore still has swapAt",
        "moveDish": "WeekStore still has moveDish",
    }
    for needle, message in banned_store.items():
        if needle in store_text:
            raise SystemExit(message)


def default_pool_path() -> Path:
    return Path(__file__).resolve().parent.parent / "KetoKasse" / "Features" / "Week" / "DishPool.swift"
